### PR TITLE
Make key of the king's law not hurt nonliving entities

### PR DESCRIPTION
**Pull request by zacharybarbanell**
Data gen has not been run on this PR.  Also includes a new custom death message for the key.

### CODE REVIEW

File: Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java [hunk 1]
```diff
@@ -33,6 +36,10 @@ private void checkInvulnerabilities(DamageSource source, CallbackInfoReturnable<
 			} else if (CrimsonPendantItem.onDamage(self, source)) {
 				cir.setReturnValue(true);
 			}
+		} else {
+			if (source.is(BotaniaTags.DamageTypes.NONLIVING_IMMUNE)) {
+				cir.setReturnValue(true);
+			}
 		}
```
Comment (williewillus) [inline Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java]: As mentioned on chat, I don't think a NONLIVING_IMMUNE tag is the right solution.

Let's invert this and scope down:
```suggestion
		}
		if (source.is(BotaniaDamageTypes.KEY_EXPLOSION) && ((Object) this) instanceof ItemEntity) { // or if you'd like, a botania:key_explosion_immune entity tag, like vanilla's minecraft:fall_damage_immune entity tag
		  cir.setReturnValue(true);
		}
```

**Review by williewillus — CHANGES_REQUESTED** [review]

File: Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java (hunk 1 above)
Comment (zacharybarbanell) [reply Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java]: Switched to two tags, one for IS_KEY (similar to vanilla's IS_FALL) and one for key_immune (similar to fall_damage_immune)

**Comment by williewillus** [comment]
Can you rebase onto 1.20, squash, and remove the is_key tag? I don't think it's necessary.

**Comment by williewillus** [comment]
Otherwise this looks good

File: Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java [hunk 2]
```diff
@@ -10,20 +10,29 @@
...
-public class EntityMixin {
+public abstract class EntityMixin {
+
+	@Shadow
+	abstract EntityType<?> getType();
+
 	/**
 	 * Cancels some invulnerabilities conferred by items
+	 * 
+	 * Adds invulnerability to nonliving entities for #nonliving_immune
```
Comment (williewillus) [inline Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java]: update or remove

File: Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java [hunk 3]
```diff
@@ -34,6 +43,9 @@ private void checkInvulnerabilities(DamageSource source, CallbackInfoReturnable<
 				cir.setReturnValue(true);
 			}
 		}
+		if (source.is(BotaniaTags.DamageTypes.IS_KEY) && this.getType().is(BotaniaTags.Entities.KEY_IMMUNE)) {
```
Comment (williewillus) [inline Xplat/src/main/java/vazkii/botania/mixin/EntityMixin.java]: as discussed, remove IS_KEY

**Review by williewillus — CHANGES_REQUESTED** [review]

**Pull request merged**